File: 0198-house-robber/0198-house-robber.java
class Solution {
    
    int []dp;
    // public int recur(int []nums,int i){
    //     if(i>=nums.length)return 0;
    //     if(dp[i]!=-1)return dp[i];
    //     return dp[i]=Math.max(nums[i]+recur(nums,i+2),recur(nums,i+1));
    // }
    
    public int rob(int[] nums) {
        int n=nums.length;
        dp=new int[n+2];
        // Arrays.fill(dp,-1);
        // return recur(nums,0);
        
        dp[n]=dp[n+1]=0;
        
        for(int i=n-1;i>=0;i--){
            dp[i]=Math.max(nums[i]+dp[i+2],dp[i+1]);
        }
        return dp[0];
        
        
    }
}

File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
//     int dp[101]={[0 ... 100]=-1};
//     int recur(vector<int>&nums,int idx){
        
//         if(idx>=nums.size())return 0;
        
//         if(dp[idx]!=-1)return dp[idx];
        
//         return dp[idx]= max(nums[idx]+recur(nums,idx+2),recur(nums,idx+1));
        
//     }
    int dp[101]={[0 ... 100]=-1};
    int recur(int idx,vector<int>&nums){
        if(idx>=nums.size())return 0;
        if(dp[idx]!=-1)return dp[idx];
        int ans=0;
        ans=nums[idx]+recur(idx+2,nums);
        
        ans=max(nums[idx]+recur(idx+2,nums),recur(idx+1,nums));
        return dp[idx]=ans;
    }
    
    
    int rob(vector<int>& nums) {
        // if(nums.size()==0)return 0;
        // int n1=0,n2=0;
        // for(int n:nums){
        //     int temp=n1;
        //     n1=max(n2+n,n1);
        //     n2=temp;
        // }
        // return n1;
        
        // return recur(nums,0);
        
//         vector<int>dp(nums.size()+2,0);
        
        
//         for(int i=nums.size()-1;i>=0;i--){
//             dp[i]=max(dp[i+1],nums[i]+dp[i+2]);
//         }
//         return dp[0];
        return recur(0,nums);
        
        
    }
};